lnwire/node_announcement_2.go

```diff
@@ -0,0 +1,494 @@
+// io.Reader observing the specified protocol version.
+//
+// This is part of the lnwire.Message interface.
+func (n *NodeAnnouncement2) Decode(r io.Reader, _ uint32) error {
+	var (
+		color = tlv.ZeroRecordT[tlv.TlvType1, Color]()
+		alias = tlv.ZeroRecordT[tlv.TlvType3, []byte]()
+		ipv4  = tlv.ZeroRecordT[tlv.TlvType5, IPV4Addrs]()
+		ipv6  = tlv.ZeroRecordT[tlv.TlvType7, IPV6Addrs]()
+		torV3 = tlv.ZeroRecordT[tlv.TlvType9, TorV3Addrs]()
+		dns   = tlv.ZeroRecordT[tlv.TlvType11, DNSAddress]()
+	)
+	stream, err := tlv.NewStream(ProduceRecordsSorted(
+		&n.Features,
+		&n.BlockHeight,
+		&n.NodeID,
+		&n.Signature,
+		&alias,
+		&color,
+		&ipv4,
+		&ipv6,
+		&torV3,
+		&dns,
+	)...)
+	if err != nil {
+		return err
+	}
+	n.Signature.Val.ForceSchnorr()
+
+	typeMap, err := stream.DecodeWithParsedTypesP2P(r)
+	if err != nil {
+		return err
+	}
+
+	if _, ok := typeMap[n.Alias.TlvType()]; ok {
+		n.Alias = tlv.SomeRecordT(alias)
+	}
+
+	if _, ok := typeMap[n.Color.TlvType()]; ok {
+		n.Color = tlv.SomeRecordT(color)
+	}
+
+	if _, ok := typeMap[n.IPV4Addrs.TlvType()]; ok {
+		n.IPV4Addrs = tlv.SomeRecordT(ipv4)
+	}
+
+	if _, ok := typeMap[n.IPV6Addrs.TlvType()]; ok {
+		n.IPV6Addrs = tlv.SomeRecordT(ipv6)
+	}
+
+	if _, ok := typeMap[n.TorV3Addrs.TlvType()]; ok {
+		n.TorV3Addrs = tlv.SomeRecordT(torV3)
+	}
+
+	if _, ok := typeMap[n.DNSHostName.TlvType()]; ok {
+		n.DNSHostName = tlv.SomeRecordT(dns)
+	}
+
+	n.ExtraSignedFields = ExtraSignedFieldsFromTypeMap(typeMap)
+
+	return nil
+}
+
+// Encode serializes the target ChannelUpdate2 into the passed io.Writer
+// observing the protocol version specified.
+//
+// This is part of the lnwire.Message interface.
+func (n *NodeAnnouncement2) Encode(w *bytes.Buffer, _ uint32) error {
+	return EncodePureTLVMessage(n, w)
+}
+
+// MsgType returns the integer uniquely identifying this message type on the
+// wire.
+//
+// This is part of the lnwire.Message interface.
+func (n *NodeAnnouncement2) MsgType() MessageType {
+	return MsgNodeAnnouncement2
+}
+
+// A compile-time check to ensure NodeAnnouncement2 implements the Message
+// interface.
+var _ Message = (*NodeAnnouncement2)(nil)
+
+// A compile-time check to ensure NodeAnnouncement2 implements the
+// PureTLVMessage interface.
+var _ PureTLVMessage = (*NodeAnnouncement2)(nil)
+
+// A compile time check to ensure ChannelAnnouncement2 implements the
+// lnwire.SizeableMessage interface.
+var _ SizeableMessage = (*NodeAnnouncement2)(nil)
+
+// Color is a custom type to represent a color.RGBA that can be encoded/decoded
+// as a TLV record.
+type Color color.RGBA
+
+// Record returns a TLV record for encoding/decoding Color.
+func (c *Color) Record() tlv.Record {
+	return tlv.MakeStaticRecord(0, c, 3, rgbEncoder, rgbDecoder)
+}
+
+// rgbEncoder encodes a Color as RGB bytes.
+func rgbEncoder(w io.Writer, val interface{}, _ *[8]byte) error {
+	if v, ok := val.(*Color); ok {
+		buf := bytes.NewBuffer(nil)
+		err := WriteColorRGBA(buf, color.RGBA(*v))
```

Comment:
I think we can just pass `w` here? why creating the `buf`?

Reply:
WriteColorRGBA takes a buffer pointer